### PR TITLE
perl 5.26.0

### PR DESCRIPTION
**Pull request by DomT4**
Why bother.

### CODE REVIEW

**Comment by DomT4** [comment]
Rebased after https://github.com/Homebrew/homebrew-core/pull/14291 was merged.

File: Formula/perl.rb [hunk 1]
```diff
@@ -43,6 +30,10 @@ def install
       end
     end
 
+    # We don't want the build process to go snooping around externally
+    # for things like the `Test` perl module.
+    ENV.delete("PERL5LIB")
```
Comment (ilovezfs) [inline Formula/perl.rb]: I think this can wait for the ENV filtering. I don't think we need to pepper formulae with it.

File: Formula/perl.rb [hunk 2]
```diff
@@ -81,12 +72,13 @@ def caveats; <<-EOS.undent
 
     You can set that up like this:
       PERL_MM_OPT="INSTALL_BASE=$HOME/perl5" cpan local::lib
-      echo 'eval "$(perl -I$HOME/perl5/lib/perl5 -Mlocal::lib)"' >> #{shell_profile}
+      echo 'eval "$(perl -I$HOME/perl5/lib/perl5 -Mlocal::lib)"' >> #{Utils::Shell.profile}
```
Comment (ilovezfs) [inline Formula/perl.rb]: Why is this change needed?

File: Formula/perl.rb (hunk 1 above)
Comment (DomT4) [inline Formula/perl.rb]: I mean, it literally broke my local build 3 times before I worked out what was going on.

File: Formula/perl.rb (hunk 2 above)
Comment (DomT4) [inline Formula/perl.rb]: IIRC `shell_profile` is informally deprecated?

Comment (DomT4) [inline Formula/perl.rb]: [It seems to be pending](https://github.com/Homebrew/brew/blob/744e7dd1907aaf6493e6e4622a5997b562bfbf17/Library/Homebrew/compat/utils.rb#L2).

File: Formula/perl.rb (hunk 1 above)
Comment (ilovezfs) [inline Formula/perl.rb]: Then I'd suggest setting HOMEBREW_ENV_FILTERING=1

File: Formula/perl.rb (hunk 2 above)
Comment (ilovezfs) [inline Formula/perl.rb]: Reference?

Comment (DomT4) [inline Formula/perl.rb]: Refresh the page 😉

Comment (ilovezfs) [inline Formula/perl.rb]: I'd leave it as is for now as it's unrelated to the PR.

File: Formula/perl.rb [hunk 3]
```diff
@@ -81,12 +72,13 @@ def caveats; <<-EOS.undent
...
     EOS
   end
 
   test do
-    (testpath/"test.pl").write "print 'Perl is not an acronym, but JAPH is a Perl acronym!';"
-    system "#{bin}/perl", "test.pl"
```
Comment (ilovezfs) [inline Formula/perl.rb]: Existing test seems fine.

File: Formula/perl.rb (hunk 1 above)
Comment (DomT4) [inline Formula/perl.rb]: That's fine for me. Probably slightly confusing for users. It's not like `PERL5LIB` is an uncommon variable if you use `perl` even a little bit. Homebrew has probably been saved hassle by the whole bottle pouring aspect but this bites if you build from source, which yes, I know is discouraged for standard users.

File: Formula/perl.rb (hunk 2 above)
Comment (DomT4) [inline Formula/perl.rb]: If the change is coming it seems reasonable to scoop it up before it actually does become a shouty warning, but whatever makes you happy.

File: Formula/perl.rb (hunk 1 above)
Comment (ilovezfs) [inline Formula/perl.rb]: As no one else has reported this, I don't think it's needed.

File: Formula/rex.rb [hunk 4]
```diff
@@ -11,7 +12,14 @@ class Rex < Formula
     sha256 "d02002fba499b2a3ed5271ccf1c3ddbaf75629d747aed289a4f53de7aad9d081" => :yosemite
   end
 
-  depends_on :perl => "5.16"
```
Comment (ilovezfs) [inline Formula/rex.rb]: Can we not just set an appropriate new version here?

File: Formula/rex.rb [hunk 5]
```diff
@@ -206,9 +214,16 @@ class Rex < Formula
   end
 
   def install
+    # First prevent trying to sniff around elsewhere for required modules.
+    ENV.delete("PERL5LIB")
```
Comment (ilovezfs) [inline Formula/rex.rb]: Let's remove this.

File: Formula/rex.rb [hunk 6]
```diff
@@ -206,9 +214,16 @@ class Rex < Formula
...
+
     ENV.prepend_create_path "PERL5LIB", libexec/"lib/perl5"
     ENV.prepend_path "PERL5LIB", libexec/"lib"
 
+    # If you're using a new enough perl it'll blow up & require your
+    # confirmation that you know what you're doing. Skip that.
+    ENV["PERL_CANARY_STABILITY_NOPROMPT"] = "1"
```
Comment (ilovezfs) [inline Formula/rex.rb]: This seems like it needs to be reported

File: Formula/rex.rb (hunk 4 above)
Comment (DomT4) [inline Formula/rex.rb]: There's no "maximum perl version" special handling, and upstream hasn't added support for the newest version, and it causes problems.

Theoretically under Homebrew's don't unnecessarily dupe the system dance it really shouldn't be using our Perl by default anyway, although I'm aware we intentionally violate this rule in places.

File: Formula/rex.rb [hunk 7]
```diff
@@ -217,22 +232,23 @@ def install
...
-      chmod 0755, libexec/cmd
+      chmod 0555, libexec/cmd
       (bin/cmd).write_env_script(libexec/cmd, :PERL5LIB => ENV["PERL5LIB"])
       man1.install "blib/man1/#{cmd}.1"
     end
   end
 
   test do
-    assert_match "\(R\)\?ex #{version}", shell_output("#{bin}/rex -v"), "rex -v is expected to print out Rex version"
+    assert_match "\(R\)\?ex #{version}", shell_output("#{bin}/rex -v"),
```
Comment (ilovezfs) [inline Formula/rex.rb]: I'd leave it as is.

File: Formula/rex.rb [hunk 8]
```diff
@@ -217,22 +232,23 @@ def install
...
+      "rex -v is expected to print out Rex version"
     system bin/"rexify", "brewtest"
-    assert (testpath/"brewtest/Rexfile").exist?, "rexify is expected to create a new Rex project and pre-populate its Rexfile"
+    assert_predicate testpath/"brewtest/Rexfile", :exist?,
```
Comment (ilovezfs) [inline Formula/rex.rb]: same.

File: Formula/rex.rb (hunk 7 above)
Comment (DomT4) [inline Formula/rex.rb]: It's way, way past 80 characters.

File: Formula/vim@7.4.rb [hunk 9]
```diff
@@ -77,9 +77,18 @@ def install
       opts << "--without-x"
     end
 
-    if build.with? "luajit"
```
Comment (ilovezfs) [inline Formula/vim@7.4.rb]: I don't think we should be maintaining versioned formulae like this, nor is the change related to the PR.

File: Formula/rex.rb (hunk 8 above)
Comment (DomT4) [inline Formula/rex.rb]: Same as my above comment.

File: Formula/vim@7.4.rb [hunk 10]
```diff
@@ -107,18 +116,20 @@ def install
...
-
-      # Get linked framework using otool
-      otool_output = `otool -L #{vim_path} | grep -m 1 Python`.gsub(/\(.*\)/, "").strip.chomp
-
-      # Expand the link and get the python exec path
-      vim_framework_path = Pathname.new(otool_output).realpath.dirname.to_s.chomp
-      system_framework_path = `python-config --exec-prefix`.chomp
-
-      assert_equal system_framework_path, vim_framework_path
+    if build.with? "python3"
```
Comment (ilovezfs) [inline Formula/vim@7.4.rb]: Same comment.

File: Formula/rex.rb (hunk 6 above)
Comment (DomT4) [inline Formula/rex.rb]: Eh. There's a new Perl, upstream doesn't recognise it as stable yet & flags it as you're using a beta build I hope you know what the heck you're doing, etc. I can find someone to report it to but upstream are gonna basically be like "Well, yeah".

File: Formula/vim@7.4.rb (hunk 9 above)
Comment (DomT4) [inline Formula/vim@7.4.rb]: I don't understand why you'd want to keep a problematic formula problematic. That seems counterintuitive to Core not becoming a garbage fire.

File: Formula/vim@7.4.rb (hunk 10 above)
Comment (DomT4) [inline Formula/vim@7.4.rb]: And same reply.

File: Formula/perl.rb (hunk 3 above)
Comment (DomT4) [inline Formula/perl.rb]: No strong feelings either way. If it bothers you I can revert 🤷‍♂️.

File: Formula/rex.rb (hunk 7 above)
Comment (ilovezfs) [inline Formula/rex.rb]: Nonetheless I don't it makes sense to rewrite random in revision bumps over unrelated formatting issues.

File: Formula/vim@7.4.rb (hunk 9 above)
Comment (ilovezfs) [inline Formula/vim@7.4.rb]: Because the idea of versioned formulae is not to give the same level of support as the main formulae, and major rewrites like this violate that.

File: Formula/rex.rb [hunk 11]
```diff
@@ -11,7 +12,14 @@ class Rex < Formula
     sha256 "d02002fba499b2a3ed5271ccf1c3ddbaf75629d747aed289a4f53de7aad9d081" => :yosemite
   end
 
-  depends_on :perl => "5.16"
+  # This is the easiest way to go for now, until there's a new upstream
```
Comment (ilovezfs) [inline Formula/rex.rb]: This entire comment is unnecessary.

File: Formula/rex.rb (hunk 7 above)
Comment (DomT4) [inline Formula/rex.rb]: I can pull out the formatting issue from the revision bump commit if you like, but at some point it kind of seems like you're creating unnecessary style deviations for the sake of unnecessary style deviations. It's your repo to run, but Homebrew literally emphases style fixes as a potential contribution point.

File: Formula/rex.rb (hunk 6 above)
Comment (ilovezfs) [inline Formula/rex.rb]: I'd just remove it in that case.

File: Formula/rex.rb (hunk 4 above)
Comment (ilovezfs) [inline Formula/rex.rb]: If it's not flagged by `brew audit`/`brew style` it's not worth changing in a random revision bump.

File: Formula/rex.rb (hunk 7 above)
Comment (ilovezfs) [inline Formula/rex.rb]: If it's not flagged by `brew audit`/`brew style` it's not worth changing in a random revision bump.

File: Formula/vim@7.4.rb (hunk 9 above)
Comment (DomT4) [inline Formula/vim@7.4.rb]: I'm not sure who wins anything if the default position on versioned formulae is to let them rot, bugs and all. Applying the improved test is beneficial for Homebrew when this rolls through CI at any point in time, and applying this fix potentially stops users ending up with confusing linkage locally. It's not like either of these things is a nothing issue.

**Comment by DomT4** [comment]
I can't be bothered, to be honest. Someone else, have at it.

**Comment by Alhadis** [comment]
~~~perl
$/ = "WHOOO!";
my @turn = <>; # yoink
~~~

File: Formula/rex.rb (hunk 6 above)
Comment (Alhadis) [inline Formula/rex.rb]: @ilovezfs You should probably read [the exact wording shown to users](https://pastebin.com/PvAe04rL). Not only does it fill an entire screen, the build pauses until the user presses <kbd>return</kbd>, since the line asking *"Continue? [y ]"* is hidden for some reason. Really poor UX.

Moreover, I don't believe the `rex` author has anything to do with this: the Canary warning is being used by the `JSON::XS` module (which was developed by the author of the long-winded message itself).

File: Formula/rex.rb [hunk 12]
```diff
@@ -217,22 +232,23 @@ def install
 
     perl_build
     (libexec/"lib").install "blib/lib/Rex", "blib/lib/Rex.pm"
-
-    # Use system Perl
-    inreplace("bin/rexify") { |s| s.gsub!(/^#!perl$/, "#!/usr/bin/env perl") }
+    inreplace "bin/rex", "#!/usr/bin/env perl", "#!/usr/bin/perl"
```
Comment (Alhadis) [inline Formula/rex.rb]: @DomT4 What was the reason behind this? When I tried running `rex -v`, it threw up:

~~~
Usage: DynaLoader::dl_find_symbol(libhandle, symbolname) at /usr/local/Cellar/rex/1.5.0_1/libexec/lib/perl5/darwin-thread-multi-2level/XSLoader.pm line 89.
Compilation failed in require at /usr/local/Cellar/rex/1.5.0_1/libexec/lib/Rex/CLI.pm line 18.
BEGIN failed--compilation aborted at /usr/local/Cellar/rex/1.5.0_1/libexec/lib/Rex/CLI.pm line 18.
Compilation failed in require at /usr/local/Cellar/rex/1.5.0_1/libexec/rex line 21.
BEGIN failed--compilation aborted at /usr/local/Cellar/rex/1.5.0_1/libexec/rex line 21.
~~~

Which is what I usually see if I run a Perl program using the system's version...

File: Formula/rex.rb (hunk 6 above)
Comment (Alhadis) [inline Formula/rex.rb]: Okay, confirmed: the guy behind that warning is being evangelical. Have a read of the [first few entries in his FAQ](http://blog.schmorp.de/2015-06-06-stableperl-faq.html), and you'll quickly get the gist.

Also:
> _"Simply because the new set of perl 5 porters (the people maintaining the standard version of perl 5) have an abysmal track record of providing stable releases. … I plan to write something in more detail about this, using references, but at the moment, I am too depressed by this and in any case, **I have to save perl**, so stay tuned for a later article on this subject."_

lolwat.
